### ISSUE TITLE
ping claims to take zero time when used with preload

### ISSUE THREAD

**Issue by nmeyerhans** [iputils/iputils#193]
Forwarded from https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=611976 This report has been open for a long time, but the behavior is still present in master. The original report:

The 'time' value displays zero milliseconds for preloaded pings and
slightly short times for normal pings. I would expect the time taken
from when the first ping was preloaded into the outbound queue; of the
order of 36ms in this example.

    # ping -w 5 -l 3 -c 3 www.google.com
    PING www.l.google.com (209.85.143.104) 56(84) bytes of data.
    64 bytes from dy-in-f104.1e100.net (209.85.143.104): icmp_seq=1 ttl=56 time=34.5 ms
    64 bytes from dy-in-f104.1e100.net (209.85.143.104): icmp_seq=2 ttl=56 time=35.1 ms
    64 bytes from dy-in-f104.1e100.net (209.85.143.104): icmp_seq=3 ttl=56 time=35.8 ms

    --- www.l.google.com ping statistics ---
    3 packets transmitted, 3 received, 0% packet loss, time 0ms       <<===== This line == <<<<<<
    rtt min/avg/max/mdev = 34.530/35.203/35.881/0.572 ms, pipe 3

Same thing happens with ping6.

    # ping6 -w 5 -l 3 -c 3 ipv6.google.com
    PING ipv6.google.com(2a00:1450:8006::93) 56 data bytes
    64 bytes from 2a00:1450:8006::93: icmp_seq=1 ttl=57 time=27.5 ms
    64 bytes from 2a00:1450:8006::93: icmp_seq=2 ttl=57 time=28.6 ms
    64 bytes from 2a00:1450:8006::93: icmp_seq=3 ttl=57 time=30.5 ms

    --- ipv6.google.com ping statistics ---
    3 packets transmitted, 3 received, 0% packet loss, time 0ms
    rtt min/avg/max/mdev = 27.539/28.934/30.582/1.270 ms, pipe 3

**Comment by kerolasa** [comment]
I did experimental change that lead me to suspect preload and timings are probably broken more than one way. In short individual icmp_seq lines have timings allover the place (sometimes ~15 ms, other times 250ms) while the line mentioned to be zero is always in 0.50ms region.

```
diff --git a/ping_common.c b/ping_common.c
index 9fbaf9d..7d5184a 100644
--- a/ping_common.c
+++ b/ping_common.c
@@ -943,7 +943,8 @@ void finish(void)
 #endif
                printf(_(", %g%% packet loss"),
                       (float)((((long long)(ntransmitted - nreceived)) * 100.0) / ntransmitted));
-               printf(_(", time %ldms"), 1000 * tv.tv_sec + (tv.tv_usec + 500) / 1000);
+               printf(_(", time %ld.%ldms"), (1000 * tv.tv_sec) + ((tv.tv_usec + 500) / 1000),
+                       (tv.tv_usec % 1000));
        }
 
        putchar('\n');
```
Fixing this bug seems to require deep dive into timing measurements in ping. That reminds me about issues 192. Considering code is hard to understand and buggy maybe ping needs semi-complete rewrite.

**Comment by ValZapod** [comment]
@kerolasa On fast device you indeed only have 0ms. But on my tablet I can have from time to time this output.
```
dextop@localhost:~$ ping -w 5 -l 3 -c 3 www.google.com                                               
PING www.google.com(lj-in-x93.1e100.net (2a00:1450:4010:c06::93)) 56 data bytes                      
64 bytes from lj-in-x93.1e100.net (2a00:1450:4010:c06::93): icmp_seq=1 ttl=45 time=23.6 ms           
64 bytes from lj-in-x93.1e100.net (2a00:1450:4010:c06::93): icmp_seq=2 ttl=45 time=23.3 ms           
64 bytes from lj-in-x93.1e100.net (2a00:1450:4010:c06::93): icmp_seq=3 ttl=45 time=23.1 ms           
                                                                                                     
--- www.google.com ping statistics ---                                                               
3 packets transmitted, 3 received, 0% packet loss, time 1ms                                          
rtt min/avg/max/mdev = 23.178/23.404/23.676/0.205 ms, pipe 3      
```

**Comment by ValZapod** [comment]
@kerolasa 1) not  %ld.%ldms but  %ld.%03ldms 2) Obviously you will have to delete + 500 (it was created by that infamous commit https://github.com/iputils/iputils/commit/5f5a34c1b4d418ea68c27c00d3a0c173dd5560d6, so if you do not need to round with + 500 you should delete it. Actually about that commit... Are you sure that +500 +50 and +5 are not already done somewhere else in the code?

I just think it is strange that the original author did not knew how to propely round values... That is very funny)) Unfortunately I am not C guru, so you will have to figure it yourself.
```
-               printf(_(", time %ldms"), 1000 * tv.tv_sec + (tv.tv_usec + 500) / 1000);
+		printf(_(", time %ld.%03ldms"), (1000 * tv.tv_sec + tv.tv_usec / 1000), 
+                       (tv.tv_usec % 1000));
```

Yeah, also after this patch I see adecvate values...
```
root@kali:/opt/SourceCode/iputils/builddir# ./ping -w 5 -l 3 -c 6 www.google.com
PING www.google.com (74.125.131.104) 56(84) bytes of data.
64 bytes from lu-in-f104.1e100.net (74.125.131.104): icmp_seq=1 ttl=46 time=22.003 ms //this was patched
64 bytes from lu-in-f104.1e100.net (74.125.131.104): icmp_seq=2 ttl=46 time=22.338 ms
64 bytes from lu-in-f104.1e100.net (74.125.131.104): icmp_seq=3 ttl=46 time=22.760 ms
^C
--- www.google.com ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 0.254ms //and this one
rtt min/avg/max/mdev = 22.003/22.367/22.760/0.309 ms, pipe 3

```

**Comment by ValZapod** [comment]
@kerolasa Ping))

> require deep dive into timing measurements in ping. 

No))

> Considering code is hard to understand and buggy maybe ping needs semi-complete rewrite.

No, this is always a bad decision.

See prev. comment, this patch of yours is veeery wrong. 

@nmeyerhans Any comments... Please also comment on 

> Are you sure that +500 +50 and +5 are not already done somewhere else in the code?

Because I really do not like commit 5f5a34c and then fixit commit https://github.com/iputils/iputils/commit/07c3c413967ec6708b64a00bae4a7cde8ef36f40 which BTW still not in debian testing((( Though the problem is rather bad.

**Comment by kerolasa** [comment]
@ValZapod send a pull request if you to change something.

**Comment by ValZapod** [comment]
@kerolasa Okay, what you do not understand in this code? Do you understand that (a) triptime is in milliseconds, (b) there are 4 variants of rounding for >= 100 000 milliseconds >= 10 000 milliseconds >= 1 000 milliseconds and >= 0 milliseconds, (c)  +500, +50, +5 are used for right (mathematically) rounding, (d) you must use  %ld.%03ldms "03" is because you need 0s after dot, because without it it will be just 0.15 when the right one is 0.015. Just rereed printf() specs. (e) the same about "time smth"

My pull request ALWAYS WILL print milliseconds with all 3 numbers after dot without ANY ROUNDING because it is impossible unless we introduce nanoseconds presicion. Are you sure about that? Of cause it is maybe a good idea to do that... Maybe if you increase verbosity, for example)) AND again why you cannot tell me for sure that 

> Are you sure that +500 +50 and +5 are not already done somewhere else in the code?

Aren't you a maintainer or smth?

**Comment by pevik** [comment]
> > Are you sure that +500 +50 and +5 are not already done somewhere else in the code?
> 
> Aren't you a maintainer or smth?

@ValZapod you're more than welcome to do cleanup or even ping rewrite (which is needed, after various people added since project was taken in 2013 from original project after YOSHIFUJI Hideaki disappeared; IMHO ping needed some cleanup even before while being in the original project).

**Comment by nmeyerhans** [comment]
> @nmeyerhans Any comments... Please also comment on

I don't specifically have any comments on the code as it exists today; I don't understand it well enough to critique it. However, I tend to agree that a rewrite is probably not necessary for this specific case. We're going to need pretty much the same math either way, and that's the part that seems to be causing problems.

IMO the right thing to do right now would be to factor the timing calculations into a standalone function and write some tests that feed it known data with known results. That'll at least allow us to validate that the calculated timing is accurate.

Additionally, I will echo the sentiment expressed in some of the other replies. Please, if you believe that the current code is incorrect, propose fixes instead of simply stating that it is incorrect. That will be far more constructive and ultimately help the maintainers to ensure that your concerns are understood and addressed.

> Because I really do not like commit [5f5a34c](https://github.com/iputils/iputils/commit/5f5a34c1b4d418ea68c27c00d3a0c173dd5560d6) and then fixit commit [07c3c41](https://github.com/iputils/iputils/commit/07c3c413967ec6708b64a00bae4a7cde8ef36f40) which BTW still not in debian testing((( Though the problem is rather bad.

07c3c41 will enter Debian testing in a few days. It is in unstable now, and has been since 20190515 was uploaded.

**Comment by ValZapod** [comment]
> will enter Debian testing in a few days. It is in unstable now, and has been since 20190515 was uploaded.

I know. https://packages.debian.org/search?keywords=iputils-ping Do you understand that there is such thing as debian patches! I want this change in Buster! Not is unstable or testing...

> That will be far more constructive and ultimately help the maintainers to ensure that your concerns are understood and addressed.

I already gave a patch

**Comment by pevik** [comment]
@ValZapod I'm also planning to have a look at the fix, but busy atm. Please understand that people maintain open source project in their free time (which is sometimes very limited).
Generally, long term plan is 1) enhance ping testing 2) rewrite it. As code is a bit complex and we don't want to introduce bugs.

**Comment by ValZapod** [comment]
Okay, I played with it a little. 

```
root@kali:/opt/SourceCode/iputils/builddir# time ./ping -w 5 -l 3 -c 3 www.google.com
PING www.google.com(lg-in-x69.1e100.net (2a00:1450:4010:c08::69)) 56 data bytes
64 bytes from lg-in-x69.1e100.net (2a00:1450:4010:c08::69): icmp_seq=1 ttl=45 ti
me=112.103 ms //fwireshark 112.078
64 bytes from lg-in-x69.1e100.net (2a00:1450:4010:c08::69): icmp_seq=2 ttl=45 ti
me=112.005 ms //wireshark 111.992
64 bytes from lg-in-x69.1e100.net (2a00:1450:4010:c08::69): icmp_seq=3 ttl=45 ti
me=111.943 ms //wireshark 111.931

--- www.google.com ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 0.284ms
rtt min/avg/max/mdev = 111.943/112.017/112.103/0.065 ms, pipe 3

real    0m0.191s
user    0m0.000s
sys     0m0.017s
```

I have Wireshark running in parallel, and this is the result: as we can see we have DNS A and AAAA (ipv6) query (in parallel), then DNS PTR query of ipv6 address. And only then we have 3 icmpv6. And then again PTR resolution. All of this is 0.177 seconds. While time prints 0.191 s
![Безымянный](https://user-images.githubusercontent.com/31514790/62765367-a8b8f600-ba98-11e9-9df2-174feef91998.png)
 and ping prints 0.000284 seconds... Hmm... Okay, there is a mistake))
